### PR TITLE
feat: add Tensor.associative_scan (Kogge-Stone parallel prefix scan)

### PR DESCRIPTION
**Pull request by GitMonsters**
Closes #3039

## Summary

Implements `Tensor.associative_scan`, an inclusive prefix scan along a given axis using an arbitrary associative binary function — similar to `jax.lax.associative_scan`.

Given input `[a, b, c, d]` and function `fn`, produces:
`[a, fn(a,b), fn(fn(a,b),c), fn(fn(fn(a,b),c),d)]`

## Algorithm

Uses the **Kogge-Stone parallel scan**: O(n log n) work, O(log n) depth.

Each round doubles the scan distance `d`:
1. Shift the tensor left by `d` along the scan axis (pad zeros, slice to original size)
2. Apply `fn(shifted, x)` at positions `i >= d` via a mask
3. Positions `i < d` are unchanged (already correct prefix values)

After `ceil(log2(n))` rounds, every position holds the correct inclusive prefix.

## API

```python
t = Tensor([1., 2., 3., 4.])

# Cumulative sum
t.associative_scan(lambda a, b: a + b)  # [1, 3, 6, 10]

# Reverse scan
t.associative_scan(lambda a, b: a + b, reverse=True)  # [10, 9, 7, 4]

# 2D along axis=1
t2d = Tensor([[1., 2., 3.], [4., 5., 6.]])
t2d.associative_scan(lambda a, b: a * b, axis=1)  # [[1, 2, 6], [4, 20, 120]]

# Running max
t.associative_scan(lambda a, b: a.maximum(b))  # [1, 2, 3, 4]
```

## Features

- Any associative binary `fn`: addition, multiplication, max, min, logical and/or, etc.
- `axis` parameter (supports negative indexing)
- `reverse=True` for right-to-left scan
- Handles edge cases: scalar (`ndim==0`), zero-size dimensions

## Tests

6 new tests in `test/backend/test_ops.py`:
- `test_associative_scan_add` — compared to `torch.cumsum`, 1D/2D/3D
- `test_associative_scan_mul` — compared to `torch.cumprod`
- `test_associative_scan_max` — compared to `torch.cummax`
- `test_associative_scan_reverse` — reverse mode
- `test_associative_scan_zero_dim` — zero-size axes
- `test_associative_scan_scalar` — scalar tensor

### CODE REVIEW

**Comment by geohot** [comment]
DO NOT USE AI
